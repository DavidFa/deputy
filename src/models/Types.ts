// export type PeopleType  = KeyValuePair;
export type FilterType = "name" | "email" | "address" | "location" | "status"

export type PeopleEditKeyType = "email" | "address" | "location" | "dob"

export interface PeopleType {
    id: string;
    name: string;
    dob?: string;
    email: string;
    address?: string;
    location?: string;
    status?: string;
    avatarUrl?: string;
}

// export interface PeoplePageStateType {
//     filteredPeople: DataGridRows[];
// }

// export interface PeoplePageActionType {
//     type: string;
//     payload: DataGridRows[];
// }

export type DropdownOptionType<V, I> = {
    value: V;
    item: I;
}

export type DataGridColumns = {
    field: string;
    headerName: string;
    align?: string;
    sortable?: boolean;
}

export interface KeyValuePair {
    [key: string]: string;
}

export interface DataGridRows extends KeyValuePair {
}

export interface DataGridProps {
    rows: DataGridRows[];
    columns: DataGridColumns[];
    rowsPerPage?: number;
    onViewClickHandler: (personId: string) => void;
}

export interface MyKnownError {
    errorMessage: string
}