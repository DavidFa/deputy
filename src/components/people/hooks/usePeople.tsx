import React, { useCallback, useEffect, useState } from 'react'
import { useAppDispatch, useAppSelector } from '../../../hooks/hooks';
import { FilterType, PeopleEditKeyType, PeopleType } from '../../../models/Types';
import { actions } from '../../../store/reducers/peopleReducer';
import { fetchPeople } from '../../../store/thunks/peopleThunks';

const usePeople = () => {
    // people fetched by API
    const { people, isLoading, isInitial } = useAppSelector(state => state.people);
    const dispatch = useAppDispatch();
    // corresponding people list shown in the table
    const [filteredPeople, setFilteredPeople] = useState<PeopleType[]>([]);
    // filter client entered
    const [filter, setFilter] = useState<string>('');
    // filter type 
    const [filterBy, setFilterBy] = useState<FilterType | ''>('');
    // person need to be edited
    const [personEdit, setPersonEdit] = useState<PeopleType | undefined>(undefined);

    // only fetch data when the page loads at first time
    useEffect(() => {
        if (isInitial)
            dispatch(fetchPeople(10));
    }, [isInitial, dispatch])

    const getFilteredPeople = useCallback(
        () => {
            const newPeople = people.filter(item => {
                if (!filter.trim()) {
                    return item;
                }
                if (filterBy) {
                    return !filter.trim() || item[filterBy]?.toLowerCase().match(filter.trim().toLowerCase())
                }
                for (const [_, value] of Object.entries(item)) {
                    if (value?.toString().toLowerCase().match(filter.trim().toLowerCase())) return item;
                }
                return false;
            });
            setFilteredPeople([...newPeople]);
        },
        [filter, filterBy, people],
    )

    // Implement debounce, alternatively we can use lodash library
    useEffect(() => {
        const timeout = setTimeout(() => { getFilteredPeople(); }, 300);
        return () => {
            clearTimeout(timeout);
        }
    }, [filter, getFilteredPeople])

    useEffect(() => {
        getFilteredPeople();
    }, [filterBy, people, getFilteredPeople])

    // useCallback will return a memoized callback that only changes if one of the dependencies has changed
    // Taking advantage of conbination of useCallback and React.memo to prevent unnecessery re-render
    const onDropdownChangeHandler = useCallback(
        (event: React.FormEvent<HTMLSelectElement>): void => {
            const filterBy: FilterType = event.currentTarget.value as FilterType;
            setFilterBy(filterBy);
        },
        [],
    )

    // useCallback will return a memoized callback that only changes if one of the dependencies has changed
    // Taking advantage of conbination of useCallback and React.memo to prevent unnecessery re-render
    const onSearchBarChangeHandler = useCallback(
        (event: React.FormEvent<HTMLInputElement>): void => {
            const filter: string = event.currentTarget.value;
            setFilter(filter);
        },
        [],
    )

    const onViewClickHandler = useCallback(
        (personId: string): void => {
            const personEdit = people.find(person => {
                return person.id === personId
            });
            setPersonEdit(personEdit);
        },
        [people],
    )

    const onPeopleEditChangeHandler = useCallback(
        (event: React.FormEvent<HTMLInputElement>, key: PeopleEditKeyType): void => {
            const value = event.currentTarget.value;
            const newPeoson = { ...personEdit } as PeopleType;
            newPeoson[key] = value;
            setPersonEdit(newPeoson);
        },
        [personEdit],
    )

    // this is only an example of validation to validate if email is empty
    const validate = useCallback(
        (): boolean => {
            if (!personEdit?.email) {
                alert("Email is empty!");
                return false
            }
            return true;
        },
        [personEdit],
    )

    const onPeopleEditFormSubmitHandler = useCallback(
        (event: React.FormEvent<HTMLFormElement>): void => {
            event.preventDefault();
            if (validate()) {
                dispatch(actions.update(personEdit as PeopleType));
                setPersonEdit(undefined);
            }
        },
        [personEdit, validate, dispatch],
    )

    return { personEdit, people, isLoading, isInitial, filterBy, filter, filteredPeople, validate, onSearchBarChangeHandler, onDropdownChangeHandler, onViewClickHandler, onPeopleEditChangeHandler, onPeopleEditFormSubmitHandler }
}

export default usePeople
