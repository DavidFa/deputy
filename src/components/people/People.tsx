import styled from 'styled-components';
import { DataGridColumns, DataGridRows, DropdownOptionType } from '../../models/Types';
import Dropdown from '../UI/Dropdown/Dropdown';
import SearchToolBar from '../UI/SearchToolbar/SearchToolbar';
import DataGrid from '../UI/Table/DataGrid';
import usePeople from './hooks/usePeople';
import PersonEdit from './PersonEdit';

const Wrapper = styled.div`
width:100%;
`;

const PageTitle = styled.h2`
float:left;
width:100%;
text-align:left
`

const filterOptions: DropdownOptionType<string, string>[] = [
    { value: "name", item: "Name" },
    { value: "email", item: "Email" },
    { value: "address", item: "Address" },
    { value: "status", item: "Status" }
];

const peopleTableHeader: DataGridColumns[] = [
    { field: 'name', headerName: 'Name' },
    { field: 'address', headerName: 'Address' },
    { field: 'location', headerName: 'Location' },
    { field: 'email', headerName: 'Email' },
    { field: 'status', headerName: 'Status' },
    { field: 'action', headerName: '' },
];

const People = () => {

    const { personEdit, isLoading, filterBy, filter, filteredPeople,
        onSearchBarChangeHandler, onDropdownChangeHandler, onViewClickHandler, onPeopleEditChangeHandler, onPeopleEditFormSubmitHandler } = usePeople();

    return (
        <Wrapper>
            {personEdit && <PersonEdit person={personEdit} onChange={onPeopleEditChangeHandler} onSubmit={onPeopleEditFormSubmitHandler} />}
            {/* Here should make a Spinner UI component for the Loading notification, but I just keep it simply as below */}
            {isLoading && <div data-testid="loading">Loading ...</div>}
            <PageTitle data-testid="pageTitle">People</PageTitle>
            <SearchToolBar placeholder="Search People..." onChangeHandler={onSearchBarChangeHandler} value={filter} />
            <Dropdown options={filterOptions} onChangeHandler={onDropdownChangeHandler} value={filterBy} />
            <DataGrid columns={peopleTableHeader} rows={filteredPeople as unknown as DataGridRows[]} onViewClickHandler={onViewClickHandler} />
        </Wrapper>
    )
}

export default People;
