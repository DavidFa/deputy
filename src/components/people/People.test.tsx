import People from './People';
import { render, screen, fireEvent } from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import * as useReduxState from '../../hooks/hooks';
import renderer from 'react-test-renderer';
import { PeopleType } from '../../models/Types';
import userEvent from '@testing-library/user-event';

describe('Testing People Component', () => {

    const mockUseDispatch = jest.fn();
    const setState = jest.fn();
    let peopleMock: PeopleType[] = [];

    const generateMockPerson = (i: number): PeopleType => {
        return { "name": `Name ${i}`, "email": `Email_${i}@gmail.com`, "address": `address_${i}`, "status": i % 2 == 1 ? "Employed" : "Discarded", "id": `${i}`, "dob": "", "location": `location_${i}`, "avatarUrl": "" };
    }

    beforeEach(() => {
        jest.spyOn(ReactRedux, 'useDispatch').mockReturnValue(mockUseDispatch);
        peopleMock = [];
        for (let i = 1; i <= 20; i++) {
            peopleMock.push(generateMockPerson(i));
        }

        jest.spyOn(useReduxState, 'useAppSelector').mockReturnValue({
            people: peopleMock,
            isLoading: true,
            errorMessage: '',
            isInitial: false
        });

        render(<People />)
    })

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("should match snapshot", () => {
        const component = renderer.create(<People />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should render a loading div', () => {
        expect(screen.getByTestId(/loading/i)).toHaveTextContent("Loading ...");
    })

    it('should render a page title', () => {

        expect(screen.getByTestId("pageTitle")).toBeInTheDocument();
        expect(screen.queryByText(/People/i)).toBeInTheDocument();
        // expect(screen.queryByText(/Title/i)).toBeInTheDocument();
        // expect(screen.queryByText(/Body/i)).toBeInTheDocument();
        // expect(screen.queryByText(/Body/i)).toBeInTheDocument();
        // expect(screen.queryByTestId(/input-title/i)).toBeInTheDocument();
        // expect(screen.queryByTestId(/input-body/i)).toBeInTheDocument();

        // const cancel = screen.getByText(/Cancel/i);
        // expect(cancel).toBeInTheDocument();

        // fireEvent.click(cancel);
        // expect(mockUseDispatch).toHaveBeenCalledTimes(1);

        // const save = screen.getByText(/Save/i);
        // expect(save).toBeInTheDocument();

        // fireEvent.click(save);
        // expect(mockUseDispatch).toHaveBeenCalledTimes(2);
        // // window.alert = jsdomAlert;
    })

    it('should render a filter input', () => {
        expect(screen.getByTestId("searchToolBar")).toBeInTheDocument();
    })

    it('should render a filter dropdown list', () => {
        expect(screen.getByTestId("dropdown")).toBeInTheDocument();
        expect((screen.getByText(/Filter/i) as HTMLOptionElement).selected).toBeTruthy();
    })

    it('should render a DataGrid', () => {
        expect(screen.getByTestId("dataGrid")).toBeInTheDocument();
    })
})