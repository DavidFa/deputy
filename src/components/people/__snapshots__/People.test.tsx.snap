// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing People Component should match snapshot 1`] = `
<div
  className="sc-fKVqWL iEteF"
>
  <div
    data-testid="loading"
  >
    Loading ...
  </div>
  <h2
    className="sc-bBHxTw kpjttF"
    data-testid="pageTitle"
  >
    People
  </h2>
  <input
    className="sc-gsDKAQ bsgvEM"
    data-testid="searchToolBar"
    onChange={[Function]}
    placeholder="Search People..."
    value=""
  />
  <select
    className="sc-bdvvtL hniPQQ"
    data-testid="dropdown"
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Filter
    </option>
    <option
      value="name"
    >
      Name
    </option>
    <option
      value="email"
    >
      Email
    </option>
    <option
      value="address"
    >
      Address
    </option>
    <option
      value="status"
    >
      Status
    </option>
  </select>
  <div
    className="sc-kDTinF kvsTBR"
    data-testid="dataGrid"
  >
    <div
      className="sc-dkPtRN cchSIX"
    >
      <table
        className="sc-hKwDye WuTZu"
      >
        <thead
          className="sc-gKclnd kQaIrJ"
        >
          <tr
            className="sc-jRQBWg hfXRKW"
          >
            <td
              className="sc-eCImPb eelyZY"
            >
              Name
            </td>
            <td
              className="sc-eCImPb eelyZY"
            >
              Address
            </td>
            <td
              className="sc-eCImPb eelyZY"
            >
              Location
            </td>
            <td
              className="sc-eCImPb eelyZY"
            >
              Email
            </td>
            <td
              className="sc-eCImPb eelyZY"
            >
              Status
            </td>
            <td
              className="sc-eCImPb eelyZY"
            >
              
            </td>
          </tr>
        </thead>
        <tbody
          className="sc-iCfMLu dkvyUx"
        >
          <tr
            className="sc-jRQBWg hfXRKW"
          >
            <td
              className="sc-eCImPb exNnNj"
              colSpan={6}
            >
              Data is empty!
            </td>
          </tr>
        </tbody>
        <tfoot
          className="sc-pVTFL cSzJGF"
        >
          <tr
            className="sc-jRQBWg hfXRKW"
          >
            <td
              className="sc-eCImPb hvhnER"
            >
              <button
                className="sc-furwcr jBqzNd"
                disabled={true}
                onClick={[Function]}
              >
                Previous Page
              </button>
            </td>
            <td
              className="sc-eCImPb eUVqQa"
              colSpan={4}
            >
              <div
                className="sc-jrQzAO jwrtrC"
              />
            </td>
            <td
              className="sc-eCImPb eUVqQa"
            >
              <button
                className="sc-furwcr jBqzNd"
                disabled={false}
                onClick={[Function]}
              >
                Next Page
              </button>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</div>
`;
